pkg/client/client.go

```diff
@@ -761,6 +763,126 @@ func (c *Client) WaitForPrometheus(ctx context.Context, p *monv1.Prometheus) err
 	return nil
 }
 
+// ValidationError indicates if an operation performed by the client has
+// resulted in an invalid state.
+type ValidationError struct {
+	Degraded    bool
+	Unavailable bool
+	Reasons     []string
+}
+
+func (ve *ValidationError) Error() string {
+	return fmt.Sprintf("error: %s degraded: %t unavailable: %t",
+		strings.Join(ve.Reasons, ";"), ve.Degraded, ve.Unavailable)
+}
+
+var _ error = (*ValidationError)(nil)
+
+func (c *Client) validatePromethuesPodState(ctx context.Context, p *monv1.Prometheus) *ValidationError {
+
+	status, _, err := prometheusoperator.Status(ctx, c.kclient.(*kubernetes.Clientset), p)
+	if err != nil {
+		klog.V(4).ErrorS(err, "validatePrometheusPodState: failed to get Prometheus status")
+		// TODO(sthaha): do we want to not set Degraded in this case?
+		return &ValidationError{Degraded: true, Reasons: []string{"failed to get Prometheus status"}}
+	}
+
+	// NOTE: about status (which seems a bit confusing)
+	// Available       = pods in Running && Ready  **but** MAY NOT be up to date
+	// UpdatedReplicas = Pods are Running && Ready && matches Spec
+	// Unavailable		 = Pods that not Available; i.e. any state other than Running & Ready
+	//
+	// Total Pods = Available + Unavailable
+	// Pods Running but NEEDS updation = Available - UpdatedReplicas (also returned as []pods)
+	// NeedsUpdation = Available - UpdatedReplicas
+
+	required := *p.Spec.Replicas
+	available := status.AvailableReplicas
+	uptoDate := status.UpdatedReplicas
+
+	needsUpdation := available - uptoDate
+
+	if uptoDate >= required || needsUpdation >= required {
+		// All pods are upto date and nothing to do; see reason above
+		//   -- OR --
+		// all pods need to be updated but they are running, in which case, we
+		// treat it as valid and the next reconciliation loop will tell us if the
+		// states are okay or not
+		return nil
+	}
+
+	// some or all  pods are NOT running, find the reason
+	// FIXME(sthaha): this assumes that the p.Spec.Shard is nil which is the case in OCP
```

Comment:
@jan--f Fixing this requires upstream change to promoperator. Which I will so that can call `promoperator.Pods(prometheus)`

Reply:
https://github.com/prometheus-operator/prometheus-operator/pull/4580 will make the upstream fix obsolete.